src/thread_task.rs

```diff
@@ -718,92 +745,94 @@ unsafe fn delayed_fg_task(c: *const Rav1dContext, ttd: *mut TaskThreadData) {
                         &(*((*c).dsp).as_ptr().offset(off as isize)).fg,
                         &mut *out,
                         &*in_0,
-                        BitDepth16::select_mut(&mut (*ttd).delayed_fg.grain),
+                        BitDepth16::select_mut(&mut ttd_locked.delayed_fg.grain),
                     );
                 }
                 _ => {
                     abort();
                 }
             }
-            (*ttd).delayed_fg.type_0 = RAV1D_TASK_TYPE_FG_APPLY;
-            pthread_mutex_lock(&mut (*ttd).lock);
-            (*ttd).delayed_fg.exec = 1 as c_int;
+            ttd_locked.delayed_fg.type_0 = RAV1D_TASK_TYPE_FG_APPLY;
+            ttd_locked.delayed_fg.exec = 1 as c_int;
         }
         12 => {}
         _ => {
             abort();
         }
     }
     row = ::core::intrinsics::atomic_xadd_seqcst(
-        &mut *((*ttd).delayed_fg.progress).as_mut_ptr().offset(0) as *mut atomic_int,
+        &mut *((*ttd).delayed_fg_progress).as_mut_ptr().offset(0) as *mut atomic_int,
         1 as c_int,
     );
-    pthread_mutex_unlock(&mut (*ttd).lock);
+    *thread_lock = None;
     progmax = (*out).p.h + 31 >> 5;
     loop {
         if (row + 1) < progmax {
-            pthread_cond_signal(&mut (*ttd).cond);
+            (*ttd).cond.notify_one();
         } else if row + 1 >= progmax {
-            pthread_mutex_lock(&mut (*ttd).lock);
-            (*ttd).delayed_fg.exec = 0 as c_int;
+            *thread_lock = Some((*ttd).lock.lock().unwrap());
+            thread_lock.as_mut().unwrap().delayed_fg.exec = 0 as c_int;
             if row >= progmax {
                 break;
             }
-            pthread_mutex_unlock(&mut (*ttd).lock);
+            *thread_lock = None;
         }
-        match (*out).p.bpc {
-            #[cfg(feature = "bitdepth_8")]
-            8 => {
-                rav1d_apply_grain_row::<BitDepth8>(
-                    &(*((*c).dsp).as_ptr().offset(0)).fg,
-                    &mut *out,
-                    &*in_0,
-                    BitDepth8::select_mut(&mut (*ttd).delayed_fg.grain),
-                    row as usize,
-                );
-            }
-            #[cfg(feature = "bitdepth_16")]
-            10 | 12 => {
-                rav1d_apply_grain_row::<BitDepth16>(
-                    &(*((*c).dsp).as_ptr().offset(off as isize)).fg,
-                    &mut *out,
-                    &*in_0,
-                    BitDepth16::select_mut(&mut (*ttd).delayed_fg.grain),
-                    row as usize,
-                );
-            }
-            _ => {
-                abort();
+        {
+            let thread_data = (*ttd).lock.lock().unwrap();
```

Comment:
Why are we using a different lock here vs. `thread_lock`?

Reply:
Wanted a short local lock that wasn't present in the original code. That's cleaner here IMO so we don't have to move the lock in and out of the top-level `Option`.